### ISSUE TITLE
How to shuffle the dataset?

### ISSUE THREAD

**Issue by ShirleyLoveGIS** [NVIDIA/retinanet-examples#251]
I read the document for [torch.utils.data.distributed.DistributedSampler](https://pytorch.org/docs/stable/data.html#torch.utils.data.distributed.DistributedSampler).  
And I didn't find sampler.set_epoch() in the project.

Does it means the data loader is a sequential order in the training process?

If so, what shall i do to shuffle the dataset?

**Comment by yashnv** [comment]
There is no sequential ordering for sure. But can you confirm that in each epoch you see the same ordering of samples?

**Comment by ShirleyLoveGIS** [comment]
Yes. if i set batch = 1 and worker = 1, the order is the same. But when I use distributed data loader the order won't be strictly sequential but still follow a weak sequential order. I think that's caused by the different loading time for each worker. 

I want the order to be absolutely random. So now I only use one GPU (close the distributed training mode) and set `shuffle = True` when initial the dataloader in _data.py_.

**Comment by yashnv** [comment]
Does that solve the same ordering? On the othe rhand disetting the `sampler.set_epoch()` solve that?

**Comment by ShirleyLoveGIS** [comment]
I don't know where to add` sampler.set_epoch()` when using distributed training mode.

**Comment by litvinich** [comment]
I solve that with 2 things:
1. When world == 1 (1 GPU) we also need random shuffling. For that, you need to `__init__` of `DataIterator` class in `retinanet/data.py`:
```        
self.dataloader = data.DataLoader(
            self.dataset,
            batch_size=batch_size // world,
            sampler=self.sampler,
            collate_fn=self.dataset.collate_fn,
            num_workers=2,
            pin_memory=True,
            shuffle=(self.sampler is None)
        )
```
2. When world > 1 (DDP training) we need to set epoch for a sampler. You need to do this before each epoch in `retinanet/train.py`:
```
while iteration < iterations:
        if world > 1:
            data_iterator.sampler.set_epoch(iteration // len(data_iterator))
        ...
```

**Comment by ShirleyLoveGIS** [comment]
@litvinich Thx, that's exactly what I need!
